### PR TITLE
[33865] Subtasks board

### PR DESCRIPTION
**Pull request by oliverguenther**
Adds a subtasks (children) board.

- [x] Allow to filter values with debounced autocompletion from service
- [x] Extract all other services that cache all values into common base class
- [x] Custom header for parent work package with links and type name

https://community.openproject.com/wp/33865

### CODE REVIEW

File: frontend/src/app/modules/boards/board/add-list-modal/add-list-modal.component.ts [hunk 1]
```diff
@@ -39,11 +39,21 @@ import {BoardActionService} from "core-app/modules/boards/board/board-actions/bo
 import {HalResource} from "core-app/modules/hal/resources/hal-resource";
 import {AngularTrackingHelpers} from "core-components/angular/tracking-functions";
 import {CreateAutocompleterComponent} from "core-app/modules/common/autocomplete/create-autocompleter.component";
+import {of} from "rxjs";
+import {DebouncedRequestSwitchmap, errorNotificationHandler} from "core-app/helpers/rxjs/debounced-input-switchmap";
+import {ValueOption} from "core-app/modules/fields/edit/field-types/select-edit-field.component";
+import {HalResourceNotificationService} from "core-app/modules/hal/services/hal-resource-notification.service";
```
Comment (openprojectci) [inline frontend/src/app/modules/boards/board/add-list-modal/add-list-modal.component.ts:46]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/740bfb755960b0e0a392492314bb1b0e2e76f1c8/frontend/src/app/modules/boards/board/add-list-modal/add-list-modal.component.ts#L46"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

File: frontend/src/app/modules/boards/boards-root/boards-root.component.ts [hunk 2]
```diff
@@ -6,6 +6,7 @@ import {BoardVersionActionService} from "core-app/modules/boards/board/board-act
 import {QueryUpdatedService} from "core-app/modules/boards/board/query-updated/query-updated.service";
 import {BoardAssigneeActionService} from "core-app/modules/boards/board/board-actions/assignee/assignee-action.service";
 import {BoardSubprojectActionService} from "core-app/modules/boards/board/board-actions/subproject/subproject-action.service";
+import {BoardSubtasksActionService} from "core-app/modules/boards/board/board-actions/subtasks/board-subtasks-action.service";
```
Comment (openprojectci) [inline frontend/src/app/modules/boards/boards-root/boards-root.component.ts:10]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/740bfb755960b0e0a392492314bb1b0e2e76f1c8/frontend/src/app/modules/boards/boards-root/boards-root.component.ts#L10"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

**Comment by openprojectci** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/opf/openproject/blob/740bfb755960b0e0a392492314bb1b0e2e76f1c8/frontend/src/app/modules/boards/board/board-partitioned-page/board-list-container.component.ts#L25">frontend/src/app/modules/boards/board/board-partitioned-page/board-list-container.component.ts#L25</a> - Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by ulferts — COMMENTED** [review]
The code is looking fine and the implemented functionality works without flaws.

However I cannot approve the PR as I think that there are a whole lot of insufficiencies: 
* The autocompleter does only provide the subject without listing the ID so that it becomes absolutely impossible to differentiate equally named work packages. This might be fixed independently of this PR by our attempts to enrich the work package autocompleter but I think that we cannot release the feature until it is done.
* There is no possibility to add existing work packages to a list (and by that turn them into children)
* There are a whole bunch of error scenarios without adequate means of communicating them to and fixing them by the user:
  * Cross project relations in case the setting is deactivated (we can fix that be removing the setting)
  * Violations to the work package hierarchy, e.g. by moving a parent work package to the column defined by its child
  * Permissions (e.g. user lacking manage subtasks in the project)
* To me, the menu for adding lists is not explaining to the user the intend of it:
image

Whether we want to fix them in this PR or merge the PR and fix them afterwards but we will surely have to fix them before the release.

**Comment by lindenthal** [comment]
Hi @ulferts,
Thanks for reviewing it. I agree, we need to finish the things before the release. I opened a separate work package for the remaining tasks you raised. I am merging this for now since this makes it easier for me to do user acceptance tests once this feature is deployed on edge.
Thanks
Niels

**Pull request merged**